### PR TITLE
Feature/xmostest to pytest

### PR DESCRIPTION
**Pull request by kevinyeungxmos**

### CODE REVIEW

File: tests/Pyxsim/__init__.py [hunk 1]
```diff
@@ -0,0 +1,134 @@
+
```
Comment (mbanth) [inline tests/Pyxsim/__init__.py:1]: Needs a copyright comment and a comment referencing the license.  See the Source check failures in the Jenkins run.

File: tests/Pyxsim/pyxsim.py [hunk 2]
```diff
@@ -0,0 +1,448 @@
+# Copyright (c) 2011, XMOS Ltd, All rights reserved
```
Comment (mbanth) [inline tests/Pyxsim/pyxsim.py:1]: Needs the copyright date range updated to 2011-2021.

File: tests/Pyxsim/pyxsim.py [hunk 3]
```diff
@@ -0,0 +1,448 @@
...
+# This software is freely distributable under a derivative of the
+# University of Illinois/NCSA Open Source License posted in
+# LICENSE.txt and at <http://github.xcore.com/>
```
Comment (mbanth) [inline tests/Pyxsim/pyxsim.py:4]: I'm unsure whether this license reference should be replaced or our license checking script should be updated to allow it.  @xross, do you know why the pyxsim code was licensed under this license?

File: tests/Pyxsim/testers.py [hunk 4]
```diff
@@ -0,0 +1,114 @@
+import re
```
Comment (mbanth) [inline tests/Pyxsim/testers.py:1]: Needs a copyright comment and a comment referencing the license.  See the Source check failures in the Jenkins run.

File: tests/Pyxsim/xe.py [hunk 5]
```diff
@@ -0,0 +1,98 @@
+from Pyxsim.xmostest_subprocess import call, call_get_output
```
Comment (mbanth) [inline tests/Pyxsim/xe.py:1]: Needs a copyright comment and a comment referencing the license.  See the Source check failures in the Jenkins run.

File: tests/Pyxsim/xmostest_subprocess.py [hunk 6]
```diff
@@ -0,0 +1,250 @@
+import platform
```
Comment (mbanth) [inline tests/Pyxsim/xmostest_subprocess.py:1]: Needs a copyright comment and a comment referencing the license.  See the Source check failures in the Jenkins run.

**Review by mbanth — CHANGES_REQUESTED** [review]
In addition to the copyright and license issues noted in the source files, I see that the source file check has detected two other files needing attention: test_suspend_resume/src/main.xc and usb_session.py.  I believe that @xross created these files, and I request that he correct the copyright date range and add in the necessary license reference.

The change log checker has flagged up the fact that the Change Log still shows HEAD as the version.  Now that we know the next release will be 2.0.0, could we go ahead and use it for the next change log version (and likewise and change needed in module_build_info)?

File: tests/test_bulk_loopback.py [hunk 7]
```diff
@@ -1,12 +1,12 @@
 #!/usr/bin/env python
 # Copyright 2016-2021 XMOS LIMITED.
 # This Software is subject to the terms of the XMOS Public Licence: Version 1.
-import xmostest
 from usb_packet import *
 import usb_packet
 from helpers import do_usb_test, RunUsbTest
 from usb_session import UsbSession
 from usb_transaction import UsbTransaction
+import pytest
```
Comment (DaveAtkinsonXMOS) [inline tests/test_bulk_loopback.py:9]: Is this import required? I can't see any pytest-imported code here..?

File: tests/test_bulk_rx_basic.py [hunk 8]
```diff
@@ -1,13 +1,12 @@
...
 # Copyright 2016-2021 XMOS LIMITED.
 # This Software is subject to the terms of the XMOS Public Licence: Version 1.
-
-import xmostest
 from usb_packet import *
 import usb_packet
 from helpers import do_usb_test, RunUsbTest
 from usb_session import UsbSession
 from usb_transaction import UsbTransaction
+import pytest
```
Comment (DaveAtkinsonXMOS) [inline tests/test_bulk_rx_basic.py:9]: Is this import required here?

File: tests/test_bulk_rx_basic_badcrc32.py [hunk 9]
```diff
@@ -1,15 +1,15 @@
 #!/usr/bin/env python
 # Copyright 2016-2021 XMOS LIMITED.
 # This Software is subject to the terms of the XMOS Public Licence: Version 1.
-import xmostest
 from usb_packet import *
 import usb_packet
 from helpers import do_usb_test, RunUsbTest
 from usb_session import UsbSession
 from usb_transaction import UsbTransaction
+import pytest
```
Comment (DaveAtkinsonXMOS) [inline tests/test_bulk_rx_basic_badcrc32.py:9]: Is this required?

File: tests/test_bulk_rx_basic_badpid.py [hunk 10]
```diff
@@ -1,11 +1,11 @@
 #!/usr/bin/env python
 # Copyright 2016-2021 XMOS LIMITED.
 # This Software is subject to the terms of the XMOS Public Licence: Version 1.
-import xmostest
 from usb_packet import *
 from helpers import do_usb_test, RunUsbTest
 from usb_session import UsbSession
 from usb_transaction import UsbTransaction
+import pytest
```
Comment (DaveAtkinsonXMOS) [inline tests/test_bulk_rx_basic_badpid.py:8]: Is this required?

File: tests/test_bulk_rx_basic_nodata.py [hunk 11]
```diff
@@ -3,12 +3,11 @@
 # This Software is subject to the terms of the XMOS Public Licence: Version 1.
 
 # Rx out of seq (but valid.. ) data PID
-import xmostest
 from usb_packet import TokenPacket, USB_PID
 from helpers import do_usb_test, RunUsbTest
 from usb_session import UsbSession
 from usb_transaction import UsbTransaction
-
+import pytest
```
Comment (DaveAtkinsonXMOS) [inline tests/test_bulk_rx_basic_nodata.py:10]: Is this required?

File: tests/test_bulk_rx_basic_rxerror.py [hunk 12]
```diff
@@ -1,12 +1,12 @@
 #!/usr/bin/env python
 # Copyright 2016-2021 XMOS LIMITED.
 # This Software is subject to the terms of the XMOS Public Licence: Version 1.
-import xmostest
 from usb_packet import *
 import usb_packet
 from helpers import do_usb_test, RunUsbTest
 from usb_session import UsbSession
 from usb_transaction import UsbTransaction
+import pytest
```
Comment (DaveAtkinsonXMOS) [inline tests/test_bulk_rx_basic_rxerror.py:9]: Is this required?

File: tests/test_bulk_rx_multiep.py [hunk 13]
```diff
@@ -1,13 +1,12 @@
...
 # Copyright 2019-2021 XMOS LIMITED.
 # This Software is subject to the terms of the XMOS Public Licence: Version 1.
-
-import xmostest
 from usb_packet import *
 import usb_packet
 from helpers import do_usb_test, RunUsbTest
 from usb_session import UsbSession
 from usb_transaction import UsbTransaction
+import pytest
```
Comment (DaveAtkinsonXMOS) [inline tests/test_bulk_rx_multiep.py:9]: Is this required?

File: tests/test_bulk_rx_traffic.py [hunk 14]
```diff
@@ -1,14 +1,13 @@
...
 # This Software is subject to the terms of the XMOS Public Licence: Version 1.
-
-import xmostest
 from usb_packet import *
 import usb_packet
 from helpers import do_usb_test, RunUsbTest
 from usb_session import UsbSession
 from usb_transaction import UsbTransaction
 from usb_phy import USB_MAX_EP_ADDRESS
+import pytest
```
Comment (DaveAtkinsonXMOS) [inline tests/test_bulk_rx_traffic.py:10]: Not required?

File: tests/helpers.py [hunk 15]
```diff
@@ -83,116 +100,83 @@ def get_usb_clk_phy(
...
+
+            tester_list.append(tester)
+            simargs = get_sim_args(testname, clk, phy, arch)
+            simthreads = [clk, phy] + extra_tasks
+            run_on_simulator(binary, simthreads, simargs = simargs)  
+        delete_test_specific_xn_files(testname)
+        return tester_list
+    else:
+        delete_test_specific_xn_files(testname)
+        return ["Build Failed"]
```
Comment (DaveAtkinsonXMOS) [inline tests/helpers.py:179]: Refactor for single exit point:
```
    if build_success:
        for session in sessions:
            ...
            run_on_simulator(binary, simthreads, simargs=simargs)
    else:
        test_list.append("Build Failed")

    delete_test_specific_xn_files(testname)
    return tester_list
```

File: tests/test_bulk_tx_basic.py [hunk 16]
```diff
@@ -1,11 +1,10 @@
 #!/usr/bin/env python
 # Copyright 2016-2021 XMOS LIMITED.
 # This Software is subject to the terms of the XMOS Public Licence: Version 1.
-
-import xmostest
 from helpers import do_usb_test, RunUsbTest
 from usb_session import UsbSession
 from usb_transaction import UsbTransaction
+import pytest
```
Comment (DaveAtkinsonXMOS) [inline tests/test_bulk_tx_basic.py:7]: Not required.

File: tests/test_bulk_tx_basic_short.py [hunk 17]
```diff
@@ -1,10 +1,10 @@
 #!/usr/bin/env python
 # Copyright 2016-2021 XMOS LIMITED.
 # This Software is subject to the terms of the XMOS Public Licence: Version 1.
-import xmostest
 from helpers import do_usb_test, RunUsbTest
 from usb_session import UsbSession
 from usb_transaction import UsbTransaction
+import pytest
```
Comment (DaveAtkinsonXMOS) [inline tests/test_bulk_tx_basic_short.py:7]: Not required.

File: tests/test_bulk_tx_multiep.py [hunk 18]
```diff
@@ -1,11 +1,10 @@
 #!/usr/bin/env python
 # Copyright 2016-2021 XMOS LIMITED.
 # This Software is subject to the terms of the XMOS Public Licence: Version 1.
-
-import xmostest
 from helpers import do_usb_test, RunUsbTest
 from usb_session import UsbSession
 from usb_transaction import UsbTransaction
+import pytest
```
Comment (DaveAtkinsonXMOS) [inline tests/test_bulk_tx_multiep.py:7]: Not required.

File: tests/test_bulk_tx_noack.py [hunk 19]
```diff
@@ -1,11 +1,11 @@
 #!/usr/bin/env python
 # Copyright 2016-2021 XMOS LIMITED.
 # This Software is subject to the terms of the XMOS Public Licence: Version 1.
-import xmostest
 from usb_packet import USB_PID, TokenPacket, RxDataPacket
 from helpers import do_usb_test, RunUsbTest
 from usb_session import UsbSession
 from usb_transaction import UsbTransaction
+import pytest
```
Comment (DaveAtkinsonXMOS) [inline tests/test_bulk_tx_noack.py:8]: Not required.

File: tests/test_control_basic_get.py [hunk 20]
```diff
@@ -13,6 +12,7 @@
 from helpers import do_usb_test, RunUsbTest
 from usb_session import UsbSession
 from usb_transaction import UsbTransaction
+import pytest
```
Comment (DaveAtkinsonXMOS) [inline tests/test_control_basic_get.py:15]: Not required.

File: tests/test_control_basic_set.py [hunk 21]
```diff
@@ -13,6 +12,7 @@
 from helpers import do_usb_test, RunUsbTest
 from usb_session import UsbSession
 from usb_transaction import UsbTransaction
+import pytest
```
Comment (DaveAtkinsonXMOS) [inline tests/test_control_basic_set.py:15]: Not required.

File: tests/test_invalidtoken.py [hunk 22]
```diff
@@ -13,6 +12,7 @@
 from helpers import do_usb_test, RunUsbTest
 from usb_session import UsbSession
 from usb_transaction import UsbTransaction
+import pytest
```
Comment (DaveAtkinsonXMOS) [inline tests/test_invalidtoken.py:15]: Not required.

File: tests/test_iso_loopback.py [hunk 23]
```diff
@@ -1,13 +1,12 @@
...
 # Copyright 2016-2021 XMOS LIMITED.
 # This Software is subject to the terms of the XMOS Public Licence: Version 1.
-
-import xmostest
 from usb_packet import *
 import usb_packet
 from helpers import do_usb_test, RunUsbTest
 from usb_session import UsbSession
 from usb_transaction import UsbTransaction
+import pytest
```
Comment (DaveAtkinsonXMOS) [inline tests/test_iso_loopback.py:9]: Not required.

File: tests/test_iso_rx_basic.py [hunk 24]
```diff
@@ -1,10 +1,10 @@
 #!/usr/bin/env python
 # Copyright 2016-2021 XMOS LIMITED.
 # This Software is subject to the terms of the XMOS Public Licence: Version 1.
-import xmostest
 from helpers import do_usb_test, RunUsbTest
 from usb_session import UsbSession
 from usb_transaction import UsbTransaction
+import pytest
```
Comment (DaveAtkinsonXMOS) [inline tests/test_iso_rx_basic.py:7]: Not required.

File: tests/test_iso_rxtx_fastpacket.py [hunk 25]
```diff
@@ -1,13 +1,12 @@
...
 # Copyright 2016-2021 XMOS LIMITED.
 # This Software is subject to the terms of the XMOS Public Licence: Version 1.
-
-import xmostest
 from usb_packet import *
 import usb_packet
 from helpers import do_usb_test, RunUsbTest
 from usb_session import UsbSession
 from usb_transaction import UsbTransaction
+import pytest
```
Comment (DaveAtkinsonXMOS) [inline tests/test_iso_rxtx_fastpacket.py:9]: Not required.

File: tests/test_iso_tx_basic.py [hunk 26]
```diff
@@ -1,10 +1,10 @@
 #!/usr/bin/env python
 # Copyright 2016-2021 XMOS LIMITED.
 # This Software is subject to the terms of the XMOS Public Licence: Version 1.
-import xmostest
 from helpers import do_usb_test, RunUsbTest
 from usb_session import UsbSession
 from usb_transaction import UsbTransaction
+import pytest
```
Comment (DaveAtkinsonXMOS) [inline tests/test_iso_tx_basic.py:7]: Not required.

File: tests/test_ping_rx_basic.py [hunk 27]
```diff
@@ -16,6 +14,7 @@
 from helpers import do_usb_test, RunUsbTest
 from usb_session import UsbSession
 from usb_transaction import UsbTransaction
+import pytest
```
Comment (DaveAtkinsonXMOS) [inline tests/test_ping_rx_basic.py:17]: Not required.

File: tests/test_shorttoken.py [hunk 28]
```diff
@@ -1,12 +1,11 @@
 #!/usr/bin/env python
 # Copyright 2019-2021 XMOS LIMITED.
 # This Software is subject to the terms of the XMOS Public Licence: Version 1.
-
-import xmostest
 from usb_packet import TxPacket, USB_PID
 from helpers import do_usb_test, RunUsbTest
 from usb_session import UsbSession
 from usb_transaction import UsbTransaction
+import pytest
```
Comment (DaveAtkinsonXMOS) [inline tests/test_shorttoken.py:8]: Not required.

File: tests/test_sof_badcrc.py [hunk 29]
```diff
@@ -1,12 +1,12 @@
 #!/usr/bin/env python
 # Copyright 2019-2021 XMOS LIMITED.
 # This Software is subject to the terms of the XMOS Public Licence: Version 1.
-import xmostest
 from usb_packet import *
 import usb_packet
 from helpers import do_usb_test, RunUsbTest
 from usb_session import UsbSession
 from usb_transaction import UsbTransaction
+import pytest
```
Comment (DaveAtkinsonXMOS) [inline tests/test_sof_badcrc.py:9]: Not required.

File: tests/test_sof_basic.py [hunk 30]
```diff
@@ -1,11 +1,11 @@
 #!/usr/bin/env python
 # Copyright 2019-2021 XMOS LIMITED.
 # This Software is subject to the terms of the XMOS Public Licence: Version 1.
-import xmostest
 from usb_packet import CreateSofToken
 from helpers import do_usb_test, RunUsbTest
 from usb_session import UsbSession
 from usb_transaction import UsbTransaction
+import pytest
```
Comment (DaveAtkinsonXMOS) [inline tests/test_sof_basic.py:8]: Not required.

File: tests/test_suspend_resume.py [hunk 31]
```diff
@@ -1,14 +1,13 @@
...
 # This Software is subject to the terms of the XMOS Public Licence: Version 1.
-
-import xmostest
 import usb_packet
 from usb_packet import CreateSofToken
 from helpers import do_usb_test, RunUsbTest
 from usb_session import UsbSession
 from usb_transaction import UsbTransaction
 from usb_signalling import UsbSuspend, UsbResume
+import pytest
```
Comment (DaveAtkinsonXMOS) [inline tests/test_suspend_resume.py:10]: Not required.

**Review by DaveAtkinsonXMOS — COMMENTED** [review]
Should we make it a requirement, or at least a strong suggestion, that we run our python code through a formatter such as 'black' as I believe Ross has been doing...? As an additional suggestion that I've meant to mention before, how about a static analyzer, too? Pylint and/or flake8 work for me (I use both on my own code...).

File: tests/Pyxsim/xmostest_subprocess.py [hunk 32]
```diff
@@ -0,0 +1,250 @@
...
+        # Now terminate and join the main process.  If the join has not returned
+        # within 10 seconds then we will have to forcibly kill the process group
+        process.terminate()
+        process.join(timeout=10)
+
+        if process.is_alive():
+            # If the process hasn't shutdown politely yet kill it
+            try:
+                sys.stderr.write("Sending SIGKILL to PID %d that hasn't responded to SIGINT\n" % pid)
+                os.killpg(os.getpgid(pid), signal.SIGKILL)
```
[the remark this answers is not in the document]
Comment (kevinyeungxmos) [reply tests/Pyxsim/xmostest_subprocess.py:45]: process is not an integer, this is a multiprocessing object so line 13 should be like: pid = process.pid return the pid of this process.

File: tests/helpers.py [hunk 33]
```diff
@@ -11,12 +12,28 @@
...
 from usb_packet import BusReset
-
-args = None
-
-ARCHITECTURE_CHOICES = ["xs2", "xs3"]
-BUSSPEED_CHOICES = ["FS", "HS"]
-
+import pytest
+import tempfile
+import shutil
```
[the remark this answers is not in the document]
Comment (kevinyeungxmos) [reply tests/helpers.py:17]: cool! I would run with the formatter.

File: tests/Pyxsim/pyxsim.py [hunk 34]
```diff
@@ -0,0 +1,448 @@
...
+      if d:
+        v = self._xsi.sample_port_pins(tile, p, mask)
+        if bit:
+          v = v >> bit
+        vals.append(v)
+      else:
+        vals.append(0)
+    def _resume(xsi):
+      for i in range(len(ps)):
+        (package, pin, _), (tile, p, bit, mask) = ps[i]
```
[the remark this answers is not in the document]
Comment (kevinyeungxmos) [reply tests/Pyxsim/pyxsim.py:153]: corrected

File: tests/test_ping_rx_basic.py (hunk 27 above)
Comment (kevinyeungxmos) [reply tests/test_ping_rx_basic.py:17]: removed

File: tests/Pyxsim/__init__.py (hunk 1 above)
Comment (kevinyeungxmos) [reply tests/Pyxsim/__init__.py:1]: added

File: tests/Pyxsim/pyxsim.py (hunk 3 above)
Comment (kevinyeungxmos) [reply tests/Pyxsim/pyxsim.py:4]: changed

File: tests/Pyxsim/testers.py (hunk 4 above)
Comment (kevinyeungxmos) [reply tests/Pyxsim/testers.py:1]: added

File: tests/Pyxsim/xe.py (hunk 5 above)
Comment (kevinyeungxmos) [reply tests/Pyxsim/xe.py:1]: added

File: tests/Pyxsim/xmostest_subprocess.py (hunk 6 above)
Comment (kevinyeungxmos) [reply tests/Pyxsim/xmostest_subprocess.py:1]: added

File: tests/Pyxsim/pyxsim.py (hunk 3 above)
Comment (xross) [reply tests/Pyxsim/pyxsim.py:4]: I don't know I'm afraid..

File: tests/Pyxsim/__init__.py (hunk 1 above)
Comment (xross) [reply tests/Pyxsim/__init__.py:1]: I will fix on #165 if that's acceptible

Comment (xross) [reply tests/Pyxsim/__init__.py:1]: actually these are already fixed..

File: tests/Pyxsim/xmostest_subprocess.py (hunk 32 above)
[the remark this answers is not in the document]
Comment (xross) [reply tests/Pyxsim/xmostest_subprocess.py:45]: > Should we make it a requirement, or at least a strong suggestion, that we run our python code through a formatter such as 'black' as I believe Ross has been doing...? As an additional suggestion that I've meant to mention before, how about a static analyzer, too? Pylint and/or flake8 work for me (I use both on my own code...).

Agreed. I would like to see black running in the CI

File: tests/Pyxsim/__init__.py (hunk 1 above)
Comment (mbanth) [reply tests/Pyxsim/__init__.py:1]: Cool

**Comment by xross** [comment]
As discussed, nervous about pyxsim living in lib_xud (see #79)

File: tests/Pyxsim/xmostest_subprocess.py [hunk 35]
```diff
@@ -0,0 +1,250 @@
...
+    if not re.match('.*Darwin.*', ostype) and re.match('.*[W|w]in.*', ostype):
+        return True
+    else:
+        return False
+
+if platform_is_windows():
+    concat_args = True
+    use_shell = True
+    # Windows version of Python 2.7 doesn't support SIGINT
+    SIGINT = signal.SIGTERM
```
[the remark this answers is not in the document]
Comment (kevinyeungxmos) [reply tests/Pyxsim/xmostest_subprocess.py:74]: Added a case check if it is not python3, raise a TestError.

**Comment by xross** [comment]
I have tried running, currently get two unexpected test fails:

test_bulk_rx_traffic
test_stall_basic

**Comment by xross** [comment]
@kevinyeungxmos how do I enable tracing? Previously was --trace option to runtests

**Comment by kevinyeungxmos** [comment]
> @kevinyeungxmos how do I enable tracing? Previously was --trace option to runtests

In helpers.py line 18 mark trace as True

**Comment by xross** [comment]
> > @kevinyeungxmos how do I enable tracing? Previously was --trace option to runtests
> 
> In helpers.py line 18 mark trace as True

Okay, thanks

**Review by xross — APPROVED** [review]
I've tested this and happy it performs as expected.

Still nervous about xmostest items now residing in xud.

**Review by mbanth — APPROVED** [review]

**Pull request merged**